apps/design/design.md

```diff
@@ -0,0 +1,159 @@
+boundaries.
+
+What has resulted should be a design that divides code into logical layers and splits
+the public API from the internal structure. 
+
+## General Structure
+
+Public Surface:
+```
+apps/ - Contains all our code
+  confidential/ - The confidential application API
+  public/ - The public application API
+  cache/ - The cache interface that can be implemented to provide persistence cache storage of credentials
+```
+
+Internals:
+```
+apps/
+  internal/
+    client/ - Shared package for common calls that Public and Confidential apps share
+    json/ - Our own json encoder/decoder for special needs
+    shared/ - Holds types that need to be in multiple packages and can't be moved into a single one due to import cycles
+    requests/ - The pacakge to communicate to services to get tokens
+```
+
+### Use of the Go special internal/ directory
+
+In Go, a directory called internal/ contains packages that should only be used by other packages
+rooted at the same location.
+
+This is documented here: https://golang.org/doc/go1.4#internalpackages
+
+For example, a package .../a/b/c/internal/d/e/f can be imported only by code in the directory tree rooted at .../a/b/c. It cannot be imported by code in .../a/b/g or in any other repository.
+
+We use this featurs quite liberally to make clear what is using an internal package.  For example:
+
+```
+apps/internal/base - Only can be used by packages defined at apps/
+apps/internal/base/internal/storage - Only can be use by package client
+```
+
+## Public API
+
+The public API will be encapsulated in apps/.  apps/ has 3 packages of interest to users:
+
+- public/ - This is what MSAL calls the Public Application Client (service client)
+- confidential/ - This is what MSAL calls the Confidential Application Client (service)
+- cache/ - This provides the interfaces that must be implemented to create peristent caches for any MSAL client
+
+## Internals
+
+In this section we will be talking about internal/.
+
+### JSON Handling
+
+JSON must be handled specially in our app. The basics are, if we receive fields that our
+structs do not contain, we cannot drop them.  We must send them back to the service.
+
+To handle that, we use our own custom json package that handles this.
+
+See the design at: [Design](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/internal/json/design.md)
+
+### Backend communication
+
+Communication to the backends is done via the requests/ package. requests.Token is the client
+for all communication.
+
+requests.Token communicates via REST calls that are encapsulated in the ops/ client.
+
+## Adding A Feature
+
+This is the general way to add a new feature to MSAL:
+
+- Add the REST calls to ops.REST
+- Add the higher level manipulations to requests.Token
+- Add your logic to the app/\<client\> and access the services via your requests.Token
+
+## Notable Differences To Other Clients
+
+### Confidential applications have a user ID
```

Comment:
>have a user ID [](start = 30, length = 14)

this should be removed, can be done in a new refactoring. the suggested cache key in the cache callback should handle the scenario on how to partition the cache. A confidential client outside of the obo flow doesn't really work with user id's and generally the account would be null.

Reply:
@abhidnya13 confirm you have this issue (enforce confidential client per user) and I'll remove this.

Reply:
Tracking this here: https://github.com/AzureAD/microsoft-authentication-library-for-go/issues/120

Reply:
Removed, new notes added, abhidnya13 opened bug.